### PR TITLE
Improve experience with recoverable modules for view-only users.

### PR DESCRIPTION
**Pull request by techanvil**
## Summary

Addresses issue:

- #5376 

## Relevant technical choices

- Added a check for the `recoverableModules` feature flag being enabled to avoid calling `getRecoverableModules()` when the feature is not enabled and logging an error as a result.
- Memoized the result of `getRecoverableModules` in order to help prevent an issue with unwanted rerendering.
- As discussed here https://github.com/google/site-kit-wp/issues/5376#issuecomment-1165771399, serialized `moduleSlugs` as `metadata.moduleSlug` in order to play nicely with the existing logic in `combineWidgets() -> shouldCombineAllWidgets()`.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/googlesitekit/widgets/components/WidgetRecoverableModules.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	widgetSlug,
+	moduleSlugs,
+	...props
+} ) {
+	const metadata = useMemo(
+		() => ( {
+			// Here we serialize to `moduleSlug` for compatibility with the logic in
+			// `combineWidgets()`. In future we may wish to take a less "hacky" approach.
+			// See https://github.com/google/site-kit-wp/issues/5376#issuecomment-1165771399.
+			moduleSlug: JSON.stringify( moduleSlugs.sort() ),
```
Comment (aaemnnosttv) [inline assets/js/googlesitekit/widgets/components/WidgetRecoverableModules.js:46]: We shouldn't call `sort` on the `moduleSlugs` prop directly because it will mutate the input.

Also, let's use `join` instead of stringify because the result for a module that only relies on a single module will be only the one module's slug rather than a JSON string of the array.

```suggestion
			moduleSlug: [ ...moduleSlugs ].sort().join(),
```

File: assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.test.js [hunk 2]
```diff
@@ -16,6 +16,11 @@
  * limitations under the License.
  */
 
+/**
+ * External dependencies
+ */
+import { getByText } from '@testing-library/dom';
```
Comment (aaemnnosttv) [inline assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.test.js:22]: This shouldn't be necessary as this query is returned by `render`

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 3]
```diff
@@ -100,6 +100,29 @@ const normalizeModules = memize( ( serverDefinitions, clientDefinitions ) => {
 		}, {} );
 } );
 
+/**
+ * Memoized function to build an object mapping recoverable module slugs to their corresponding
+ * module objects.
```
Comment (aaemnnosttv) [inline assets/js/googlesitekit/modules/datastore/modules.js:105]: This should trigger our `sitekit/jsdoc-third-person` rule but it looks like it isn't run on doc blocks for functions defined as variables.

Doc blocks aren't needed for such internal functions, but it doesn't hurt to keep; let's update it to satisfy the check though for consistency :)

```suggestion
 * Gets a memoized object mapping recoverable module slugs to their corresponding
 * module objects.
```

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 4]
```diff
@@ -100,6 +100,29 @@ const normalizeModules = memize( ( serverDefinitions, clientDefinitions ) => {
...
+	Object.values( modules ).reduce( ( recoverable, module ) => {
+		if ( recoverableModules.includes( module.slug ) ) {
+			return {
+				...recoverable,
+				[ module.slug ]: module,
+			};
+		}
+
+		return recoverable;
+	}, {} )
```
Comment (aaemnnosttv) [inline assets/js/googlesitekit/modules/datastore/modules.js:123]: This would be more efficient to iterate over the recoverable slugs as there will always be less of these than all the modules (simply because not all modules are shareable). We could also then check if `modules[recoverableSlug]` rather than the list search when reducing.

The sets are so small that this probably doesn't matter at all but worth mentioning.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @techanvil, this looks and works great! Just a few minor things to address and this should be good to go. Nice work!

File: assets/js/googlesitekit/modules/datastore/modules.js (hunk 4 above)
Comment (techanvil) [reply assets/js/googlesitekit/modules/datastore/modules.js:123]: Yeah, this did cross my mind too. As I was simply moving this code, I thought it might cloud the issue if I also rewrote it. Besides, as you've pointed out the sets are small, and likely to stay that way, so it would arguably be a case of premature optimization to rewrite it...

File: assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.test.js (hunk 2 above)
Comment (techanvil) [reply assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.test.js:22]: This is a deliberate choice, as I wanted to use the version of `getByText` which accepts a [container as the first argument](https://testing-library.com/docs/queries/about#using-queries), rather than the [bound version](https://testing-library.com/docs/react-testing-library/api#queries-1) returned by `render`.

File: assets/js/googlesitekit/widgets/components/WidgetRecoverableModules.js (hunk 1 above)
Comment (techanvil) [reply assets/js/googlesitekit/widgets/components/WidgetRecoverableModules.js]: Thanks, good spot on the `sort` mutation!

I've fixed that and used `join` rather than `JSON.stringify`. I joined using a comma to make the value a bit more coherent and avoid an obscure edge case...

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, I have addressed your comments :+1:

File: assets/js/googlesitekit/widgets/components/WidgetRenderer.test.js [hunk 5]
```diff
@@ -78,4 +89,36 @@ describe( 'WidgetRenderer', () => {
 
 		expect( container.firstChild ).toEqual( null );
 	} );
+
+	it( 'should output the recoverable modules component when the widget depends on a recoverable module', async () => {
+		const { getByText } = render( <WidgetRenderer slug="TestWidget" />, {
+			setupRegistry: setupRegistry( {
+				recoverableModules: [ 'search-console' ],
+			} ),
+			viewContext: VIEW_CONTEXT_DASHBOARD_VIEW_ONLY,
```
[the remark this answers is not in the document]
Comment (techanvil) [reply assets/js/googlesitekit/widgets/components/WidgetRenderer.test.js:101]: Thanks, good shout - I've added another test case.

File: assets/js/googlesitekit/widgets/components/WidgetRecoverableModules.js (hunk 1 above)
Comment (aaemnnosttv) [reply assets/js/googlesitekit/widgets/components/WidgetRecoverableModules.js]: Great thanks! `join()` uses a comma by default but good to be explicit I suppose. You can also do `[1,2,3].toString()` which will be comma-separated as well 😉

Comment (techanvil) [reply assets/js/googlesitekit/widgets/components/WidgetRecoverableModules.js]: Well, thanks for another reminder of the basics. Mutable `sort` was an oversight on my part, but I had completely forgotten `join` defaults to a comma! Regardless, I do prefer the explicit version (probably evident from the fact I had forgotten about the default), for the sake of, well, being explicit. :)

File: assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.test.js (hunk 2 above)
Comment (aaemnnosttv) [reply assets/js/googlesitekit/widgets/components/WidgetAreaRenderer.test.js:22]: Thanks! I'm aware of the difference, the part that isn't obvious is why this would be intentionally used/needed instead?

**Review by aaemnnosttv — APPROVED** [review]
Great, thanks @techanvil !
